rust/extractor/src/translate.rs

```diff
@@ -64,6 +69,77 @@ impl CrateTranslator<'_> {
         })
     }
 
+    fn emit_location_ast_ptr<T>(
+        &mut self,
+        source: ra_ap_hir::InFile<ra_ap_syntax::AstPtr<T>>,
+    ) -> Option<trap::Label>
+    where
+        T: AstNode,
+    {
+        source
+            .file_id
+            .file_id()
+            .map(|f| (f.file_id(), source))
+            .and_then(|(file_id, source)| self.emit_file(file_id).map(|data| (data, source)))
+            .and_then(|(data, source)| {
+                let range = source.value.text_range();
+                let start = data.line_index.line_col(range.start());
+                let end = data.line_index.line_col(range.end());
+                Some(self.trap.emit_location(data.label, start, end))
+            })
```

Comment:
I _think_ the `source` can be taken by capture, without passing it around in the maps like this? Also, `clippy` told me already that `.and_then(... Some(x))` can really be just `.map(... x)`
```suggestion
        source
            .file_id
            .file_id()
            .map(|f| f.file_id())
            .and_then(|file_id| self.emit_file(file_id))
            .map(|data| {
                let range = source.value.text_range();
                let start = data.line_index.line_col(range.start());
                let end = data.line_index.line_col(range.end());
                self.trap.emit_location(data.label, start, end)
            })
```

Reply:
Indeed, I think I already did that in a later commit.